app/uk/gov/hmrc/apinotificationqueue/controller/EnhancedNotificationsController.scala

```diff
@@ -61,45 +83,65 @@ class EnhancedNotificationsController @Inject()(queueService: QueueService,
             notificationIds <- queueService.get(clientId, Some(notificationStatus))
           } yield notificationIds.map(s"/notifications/$notificationStatus/" + _.notification.notificationId.toString)
 
-        notificationIdPaths.map {idPaths =>
-          val json = Json.toJson(Notifications(idPaths))
-          logger.debug(s"listing $notificationStatus notifications $json", headers.headers)
-          Ok(json)
-        }
+        generateResponse(notificationIdPaths, headers)
+    }
+  }
+
+  private def pullByConversationId(conversationId: UUID, notificationStatus: NotificationStatus.Value): Action[AnyContent] = Action.async { implicit request =>
+
+    val headers: Headers = request.headers
+    logger.info(s"listing $notificationStatus notifications by conversationId ${conversationId.toString}", headers.headers)
+    validateClientIdHeader(request.headers, s"get $notificationStatus by conversation id") match {
+      case Left(errorResponse) => Future.successful(errorResponse.XmlResult)
+      case Right(clientId) =>
+        val notificationIdPaths: Future[List[String]] =
+          for {
+            notificationIds <- queueService.getByConversationId(clientId, conversationId: UUID, notificationStatus)
+          } yield notificationIds.map(s"/notifications/$notificationStatus/" + _.notification.notificationId.toString)
+
+        generateResponse(notificationIdPaths, headers)
+    }
+  }
+
+  private def generateResponse(notificationIdPaths: Future[List[String]], headers: Headers) = {
+    notificationIdPaths.map { idPaths =>
+      val json = Json.toJson(Notifications(idPaths))
+      logger.debug(s"returning notifications $json", headers.headers)
+      Ok(json)
     }
   }
 
-  def unpulled(id: UUID): Action[AnyContent] = pull(id, Unpulled)
-  def pulled(id: UUID): Action[AnyContent] = pull(id, Pulled)
+  def unpulled(notificationId: UUID): Action[AnyContent] = pull(notificationId, Unpulled)
+  def pulled(notificationId: UUID): Action[AnyContent] = pull(notificationId, Pulled)
 
-  private def pull(id: UUID, notificationStatus: NotificationStatus.Value): Action[AnyContent] = Action.async { implicit request =>
+  private def pull(notificationId: UUID, notificationStatus: NotificationStatus.Value): Action[AnyContent] = Action.async { implicit request =>
```

Comment:
name of this method is distracting, maybe call it getList? Or something that doesn't have the word pull in the name.

Reply:
Re-jigged naming. Now all that return lists contain the word `All`.